File: algorithm/Python3_version/leetcode1292.maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold.py
class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m, n = len(mat), len(mat[0])
        maxlen = 0
        prefixsum = [[0]*n for _ in range(m)]
        prefixsum[0][0] = mat[0][0]
        for j in range(1,n):
            prefixsum[0][j] = prefixsum[0][j-1]+mat[0][j]
        for i in range(1,m):
            prefixsum[i][0] = prefixsum[i-1][0]+mat[i][0]
        for i in range(1,m):
            for j in range(1,n):
                prefixsum[i][j] = prefixsum[i-1][j]+prefixsum[i][j-1]-prefixsum[i-1][j-1]+mat[i][j]
        for i in range(m):
            for j in range(n):
                length = maxlen+1
                while length <= min(m,n):
                    ni = i+length-1
                    nj = j+length-1
                    if ni<m and nj<n and prefixsum[ni][nj]-(prefixsum[i-1][nj] if i > 0 else 0)-(prefixsum[ni][j-1] if j > 0 else 0)+(prefixsum[i-1][j-1] if i>0 and j>0 else 0) <= threshold:
                        maxlen = max(maxlen, length)
                        if maxlen == min(m,n):
                            return maxlen
                    else:
                        break
                    length += 1
        return maxlen

File: algorithm/leetcode1292.maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold.cpp
class Solution {
public:
    int maxSideLength(vector<vector<int>>& mat, int threshold) {
    	int m = mat.size(), n = mat[0].size(), i, j, maxlen = 0, len = 0;
    	vector<vector<int>> sum(m,vector<int>(n,0));
    	for(j = 0; j < n; ++j)
    		sum[0][j] = (j > 0 ? sum[0][j-1] : 0) + mat[0][j];
    	for(i = 1; i < m; ++i)
    		sum[i][0] = sum[i-1][0] + mat[i][0];
    	for(i = 1; i < m; ++i)
    		for(j = 1; j < n; ++j)
    			sum[i][j] = sum[i-1][j] + sum[i][j-1] - sum[i-1][j-1]+mat[i][j];
    	int ni, nj, sumofarea;
		for(i = 0; i < m; ++i)
			for(j = 0; j < n; ++j)
				for(len = maxlen+1; len <= min(m,n); ++len)
				{
					ni = i+len-1;
					nj = j+len-1;
					if(ni < m && nj < n && sum[ni][nj]-(i>0?sum[i-1][nj]:0)-(j>0?sum[ni][j-1]:0)
										+(i>0&&j>0 ? sum[i-1][j-1] : 0) <= threshold)
					{
                        maxlen = max(maxlen, len);
                        if(maxlen == min(m,n))
                            return maxlen;
					}
                    else
                        break;
				}
		return maxlen;
    }
};